sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala

```diff
@@ -382,6 +413,23 @@ class MicroBatchExecution(
 
     // Generate a map from each unique source to the next available offset.
     val (nextOffsets, recentOffsets) = uniqueSources.toSeq.map {
+      case (s: FakeLatestOffsetSupportsTriggerAvailableNow, limit) =>
+        val originalSource = originalSources.getOrElse(s, throw new IllegalStateException(
```

Comment:
We can also combine this with the following pattern, but that would need to change the type of `originalSource` to be more generic (`SupportsAdmissionControl` instead of `FakeLatestOffsetSupportsTriggerAvailableNow`).

Reply:
As you actually pointed out here, I no longer see the value of having the `originalSources` map. Am I missing anything? Can we just do the unwrapping transparently inside `StreamProgress` for `FakeLatestOffsetSupportsTriggerAvailableNow`?

Reply:
For that we would have to wrap sources in the logical plan... I think having the wrapped sources only in `MicroBatchExecution.uniqueSources` is a less intrusive and safer change. What do you think?

I will also remove `originalSource` mapping and get the original source directly from the wrapped source instead.

Reply:
If we use `s.source` here to extract the original source, then we can get rid of `originalSource` as well since this is the only usage.

Reply:
Thanks @HeartSaVioR! Will follow what you suggest and remove `originalSource`.